File: gandalf/registry.py
"""Discovery of quest definitions.

Every quest lives in `quests/<name>/quest.json`:

    {
      "name": "quest00",
      "aliases": ["quest00", "js-quest00"],
      "exercises": [
        {"dir": "ex00", "title": "MY-FIRST-FILE", "module": "ex00", "points": 1}
      ]
    }

`aliases` are matched against the directory name on disk, normalized (lowercase,
dashes/underscores/spaces removed), so `js-quest01`, `js_quest01` and `JSQuest01`
all resolve to the same quest.
"""

import json
import os

QUESTS_ENV = "GANDALF_QUESTS_DIR"


def normalize(name):
    return "".join(c for c in name.lower() if c.isalnum())


class Exercise(object):
    def __init__(self, quest, spec):
        self.quest = quest
        self.dir = spec["dir"]
        self.title = spec["title"]
        self.module = spec.get("module", spec["dir"])
        self.points = int(spec["points"])

    @property
    def module_path(self):
        return os.path.join(self.quest.path, self.module + ".py")


class Quest(object):
    def __init__(self, path, spec):
        self.path = path
        self.name = spec["name"]
        self.description = spec.get("description", "")
        self.aliases = set(normalize(a) for a in spec.get("aliases", []))
        self.aliases.add(normalize(self.name))
        self.exercises = [Exercise(self, e) for e in spec["exercises"]]

    def matches(self, dirname):
        return normalize(dirname) in self.aliases

    def exercise(self, dirname):
        for ex in self.exercises:
            if ex.dir == dirname:
                return ex
        return None


def quests_dir(root):
    return os.environ.get(QUESTS_ENV) or os.path.join(root, "quests")


def load_quests(root):
    base = quests_dir(root)
    quests = []
    if not os.path.isdir(base):
        return quests
    for entry in sorted(os.listdir(base)):
        manifest = os.path.join(base, entry, "quest.json")
        if not os.path.isfile(manifest):
            continue
        with open(manifest) as fh:
            spec = json.load(fh)
        quests.append(Quest(os.path.join(base, entry), spec))
    return quests


def find_quest(quests, dirname):
    for quest in quests:
        if quest.matches(dirname):
            return quest
    return None

File: lib/registry.js
"use strict";

const fs = require("fs");
const path = require("path");

const QUESTS_ENV = "GANDALF_QUESTS_DIR";

function normalize(name) {
  return name.toLowerCase().replace(/[^a-z0-9]/g, "");
}

class Exercise {
  constructor(quest, spec) {
    this.quest = quest;
    this.dir = spec.dir;
    this.title = spec.title;
    this.module = spec.module || spec.dir;
    this.points = Number(spec.points);
    this.rspec = !!spec.rspec;
  }

  get modulePath() {
    return path.join(this.quest.path, this.module + ".js");
  }
}

class Quest {
  constructor(questPath, spec) {
    this.path = questPath;
    this.name = spec.name;
    this.description = spec.description || "";
    this.aliases = new Set((spec.aliases || []).map(normalize));
    this.aliases.add(normalize(this.name));
    this.exercises = spec.exercises.map((e) => new Exercise(this, e));
  }

  matches(dirname) {
    return this.aliases.has(normalize(dirname));
  }

  exercise(dirname) {
    return this.exercises.find((e) => e.dir === dirname) || null;
  }
}

function questsDir(root) {
  return process.env[QUESTS_ENV] || path.join(root, "quests");
}

function loadQuests(root) {
  const base = questsDir(root);
  const quests = [];
  if (!fs.existsSync(base)) return quests;

  for (const entry of fs.readdirSync(base).sort()) {
    const manifest = path.join(base, entry, "quest.json");
    if (!fs.existsSync(manifest)) continue;
    const spec = JSON.parse(fs.readFileSync(manifest, "utf8"));
    quests.push(new Quest(path.join(base, entry), spec));
  }
  return quests;
}

function findQuest(quests, dirname) {
  return quests.find((q) => q.matches(dirname)) || null;
}

module.exports = { normalize, loadQuests, findQuest, Quest, Exercise };
